compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -3553,26 +3563,139 @@ private STNode parseVarDeclRhs(STNode metadata, STNode finalKeyword, STNode type
                     expr = parseActionOrExpression();
                 }
                 semicolon = parseSemicolon();
+                hasVarInit = true;
                 break;
             case SEMICOLON_TOKEN:
                 assign = STNodeFactory.createEmptyNode();
                 expr = STNodeFactory.createEmptyNode();
                 semicolon = parseSemicolon();
                 break;
             default:
-                recover(nextToken, ParserRuleContext.VAR_DECL_STMT_RHS, metadata, finalKeyword, typedBindingPattern,
-                        isModuleVar);
-                return parseVarDeclRhs(metadata, finalKeyword, typedBindingPattern, isModuleVar);
+                recover(nextToken, ParserRuleContext.VAR_DECL_STMT_RHS, metadata, isolatedFinalQualifiers,
+                        typedBindingPattern, isModuleVar);
+                return parseVarDeclRhs(metadata, isolatedFinalQualifiers, typedBindingPattern, isModuleVar);
         }
 
         endContext();
         if (isModuleVar) {
-            return STNodeFactory.createModuleVariableDeclarationNode(metadata, finalKeyword, typedBindingPattern,
-                    assign, expr, semicolon);
+            if (hasVarInit || isolatedFinalQualifiers.isEmpty()) {
+                STNode isolatedFinalQualifierNode = STNodeFactory.createNodeList(isolatedFinalQualifiers);
+                return STNodeFactory.createModuleVariableDeclarationNode(metadata, isolatedFinalQualifierNode,
+                        typedBindingPattern, assign, expr, semicolon);
+            }
+
+            // Only final qualifier is allowed for non initialized module var declaration
```

Comment:
Thing is isolatedFinalQualifiers can only have following possibilities.
`empty`, `final`, `isolated`, `final isolated` and `isolated final`. No duplicate scenarios. 

When parsing `isolated`, `final isolated` had  to be anyway validated as we should let isolated to be part of the type descriptor.
Only missing validation scenario is `isolated final`. We can simply do that also in parser as all other scenarios are already filtered. It only adds 3 lines. Shall we keep?

Reply:
Better to leave it for the semantic analyzer. Its more of a conceptual grouping.

Reply:
@MaryamZi Sorry I forgot pass this piece of information.
We need to do semantic validation for isolated-module-non-init-var-decl.
For example in following scenario we don't log an error in parser,
```bal
isolated final object{ }  obj;
```
But in the other scenarios we were forced to validate at parser as it could be a part of the immediate type descriptor.

Reply:
Noted, will add this validation. Thanks!